File: 897.Increasing-Order-Search-Tree.py
# https://leetcode.com/problems/increasing-order-search-tree/description/
#
# algorithms
# Easy (53.1%)
# Total Accepted:    2.4k
# Total Submissions: 4.5k

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def increasingBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        res = [[]]

        def resursive(node):
            if not node:
                return
            resursive(node.left)
            res[0].extend([node.val, None])
            resursive(node.right)

        resursive(root)

        return res[0][:-1]

File: 897.Increasing-Order-Search-Tree.java
// https://leetcode.com/problems/increasing-order-search-tree/
//
// algorithms
// Easy (66.11%)
// Total Accepted:    43,879
// Total Submissions: 66,373


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    static TreeNode pre, res;
    
    public TreeNode increasingBST(TreeNode root) {
        pre = null;
        res = null;
        
        recursive(root);
        
        return res;
    }
    
    public void recursive(TreeNode node) {
        if (node == null) {
            return;
        }
        
        recursive(node.left);
        if (pre == null) {
            pre = node;
            res = node;
        } else {
            pre.right = node;
            pre = node;
            node.left = null;
        }
        recursive(node.right);
    }
}